### PR TITLE
eureka-1131: handle spot instance termination

### PR DESCRIPTION
**Pull request by krutsko**

### CODE REVIEW

File: eureka-client/src/main/java/com/netflix/appinfo/AmazonInfo.java [hunk 1]
```diff
@@ -63,6 +63,8 @@
         availabilityZone("availability-zone", "placement/"),
         publicHostname("public-hostname"),
         publicIpv4("public-ipv4"),
+        spotInstanceTerminationTime("termination-time", "spot/"),
```
Comment (qiangdavidliu) [inline eureka-client/src/main/java/com/netflix/appinfo/AmazonInfo.java]: very minor nit: mind calling this `spotTerminationTime`? That aligns more with how the rest of the enums are named.

File: eureka-client/src/main/java/com/netflix/appinfo/CloudInstanceConfig.java [hunk 2]
```diff
@@ -50,7 +50,7 @@
     };
 
     private final AmazonInfoConfig amazonInfoConfig;
-    private final RefreshableAmazonInfoProvider amazonInfoHolder;
+    /* visible for testing */ RefreshableAmazonInfoProvider amazonInfoHolder;
```
Comment (qiangdavidliu) [inline eureka-client/src/main/java/com/netflix/appinfo/CloudInstanceConfig.java]: I would appreciate this be kept final to better ensure it is always correctly instantiated. An alternative for your test can be to add a new constructor that can be used to set `RefreshableAmazonInfoProvider`, with test visibility?

File: eureka-client/src/main/java/com/netflix/appinfo/ApplicationInfoManager.java [hunk 3]
```diff
@@ -219,6 +219,21 @@ public void refreshDataCenterInfoIfRequired() {
             builder.setHostName(newAddress).setIPAddr(newIp).setDataCenterInfo(config.getDataCenterInfo());
             instanceInfo.setIsDirty();
         }
+
+        String existingSpotInstanceAction  = instanceInfo.getMetadata().get(AmazonInfo.MetaDataKey.spotInstanceAction.name);
+        DataCenterInfo dataCenterInfo = config.getDataCenterInfo();
+        if (dataCenterInfo instanceof AmazonInfo) {
+            AmazonInfo amazonInfo = (AmazonInfo) dataCenterInfo;
+            String newSpotInstanceAction = amazonInfo.getMetadata().get(AmazonInfo.MetaDataKey.spotInstanceAction.name);
```
Comment (qiangdavidliu) [inline eureka-client/src/main/java/com/netflix/appinfo/ApplicationInfoManager.java]: It is unfortunate that the API for `AmazonInfo` is a bit leaky, but the preferred API to extract data is
https://github.com/Netflix/eureka/blob/f8439886921e3655a301adb149a13eaa34268074/eureka-client/src/main/java/com/netflix/appinfo/AmazonInfo.java#L299, where the enum can be used directly.

File: eureka-client/src/main/java/com/netflix/appinfo/ApplicationInfoManager.java [hunk 4]
```diff
@@ -219,6 +219,21 @@ public void refreshDataCenterInfoIfRequired() {
...
+            if (newSpotInstanceAction != null && !newSpotInstanceAction.equals(existingSpotInstanceAction)) {
+                logger.warn(String.format("The spot instance termination action changed from: %s => %s",
+                        existingSpotInstanceAction,
+                        newSpotInstanceAction));
+                InstanceInfo.Builder builder = new InstanceInfo.Builder(instanceInfo);
+                builder.setDataCenterInfo(config.getDataCenterInfo());
```
Comment (qiangdavidliu) [inline eureka-client/src/main/java/com/netflix/appinfo/ApplicationInfoManager.java]: There is duplication of code here with code on line 218/219/220, mind consolidating them? Thanks.

File: eureka-client/src/main/java/com/netflix/appinfo/ApplicationInfoManager.java [hunk 5]
```diff
@@ -219,6 +219,21 @@ public void refreshDataCenterInfoIfRequired() {
             builder.setHostName(newAddress).setIPAddr(newIp).setDataCenterInfo(config.getDataCenterInfo());
             instanceInfo.setIsDirty();
         }
+
+        String existingSpotInstanceAction  = instanceInfo.getMetadata().get(AmazonInfo.MetaDataKey.spotInstanceAction.name);
```
Comment (qiangdavidliu) [inline eureka-client/src/main/java/com/netflix/appinfo/ApplicationInfoManager.java]: (mostly due to our bad naming) I think this is the wrong metadata map to get here, you want the metadata from the instanceInfo.getDataCenterInfo(), if it is an instance of AmazonInfo.

File: eureka-client/src/main/java/com/netflix/appinfo/ApplicationInfoManager.java [hunk 6]
```diff
@@ -219,6 +219,21 @@ public void refreshDataCenterInfoIfRequired() {
...
+        DataCenterInfo dataCenterInfo = config.getDataCenterInfo();
+        if (dataCenterInfo instanceof AmazonInfo) {
+            AmazonInfo amazonInfo = (AmazonInfo) dataCenterInfo;
+            String newSpotInstanceAction = amazonInfo.getMetadata().get(AmazonInfo.MetaDataKey.spotInstanceAction.name);
+            if (newSpotInstanceAction != null && !newSpotInstanceAction.equals(existingSpotInstanceAction)) {
+                logger.warn(String.format("The spot instance termination action changed from: %s => %s",
```
Comment (qiangdavidliu) [inline eureka-client/src/main/java/com/netflix/appinfo/ApplicationInfoManager.java]: This is more you use case, is this change worth a WARN level log? Assuming spot termination actions can change often, this'll get noisy.

File: eureka-client/src/main/java/com/netflix/appinfo/CloudInstanceConfig.java (hunk 2 above)
Comment (krutsko) [reply eureka-client/src/main/java/com/netflix/appinfo/CloudInstanceConfig.java]: Fixed.

File: eureka-client/src/main/java/com/netflix/appinfo/ApplicationInfoManager.java (hunk 4 above)
Comment (krutsko) [reply eureka-client/src/main/java/com/netflix/appinfo/ApplicationInfoManager.java]: Fixed.

File: eureka-client/src/main/java/com/netflix/appinfo/ApplicationInfoManager.java (hunk 6 above)
Comment (krutsko) [reply eureka-client/src/main/java/com/netflix/appinfo/ApplicationInfoManager.java]: I have changed to INFO. If you think that DEBUG is even better here, let me know. In our case, we don't have so much spot-instances termination.

File: eureka-client/src/main/java/com/netflix/appinfo/ApplicationInfoManager.java (hunk 3 above)
Comment (krutsko) [reply eureka-client/src/main/java/com/netflix/appinfo/ApplicationInfoManager.java]: Got it. Fixed.

File: eureka-client/src/main/java/com/netflix/appinfo/ApplicationInfoManager.java (hunk 5 above)
Comment (krutsko) [reply eureka-client/src/main/java/com/netflix/appinfo/ApplicationInfoManager.java]: Right. Fixed.

File: eureka-client/src/main/java/com/netflix/appinfo/AmazonInfo.java (hunk 1 above)
Comment (krutsko) [reply eureka-client/src/main/java/com/netflix/appinfo/AmazonInfo.java]: Sure. Changed to `spotTerminationTime`.

**Comment by krutsko** [comment]
@qiangdavidliu thanks for looking into this. I have fixed all comments.

**Comment by qiangdavidliu** [comment]
Thanks @krutsko

**Pull request merged**